rising/transforms/functional/affine.py

```diff
@@ -1,15 +1,411 @@
+                batch samples
+            * a parameter per sampler per dimension
+        None will be treated as a rotation factor of 0
+    batchsize : int
+        the number of samples per batch
+    ndim : int
+        the dimensionality of the transform
+    degree : bool
+        whether the given rotation(s) are in degrees.
+        Only valid for rotation parameters, which aren't passed as full
+        transformation matrix.
+    device : torch.device, str, optional
+        the device to put the resulting tensor to. Defaults to the default
+        device
+    dtype : torch.dtype, str, optional
+        the dtype of the resulting trensor. Defaults to the default dtype
+
+    Returns
+    -------
+    torch.Tensor
+        the homogeneous transformation matrix [N, NDIM + 1, NDIM + 1], N is
+        the batch size and NDIM is the number of spatial dimensions
+
+    """
+    if rotation is None:
+        rotation = 0
+    num_rot_params = 1 if ndim == 2 else ndim
+
+    rotation = expand_scalar_param(rotation, batchsize, num_rot_params)
+    if degree:
+        rotation = deg_to_rad(rotation)
+
+    matrix_fn = create_rotation_2d if ndim == 2 else create_rotation_3d
+    sin, cos = torch.sin(rotation), torch.cos(rotation)
+    rotation_matrix = torch.stack([matrix_fn(s, c) for s, c in zip(sin, cos)]).to(
+        device=device, dtype=dtype)
+    return matrix_to_homogeneous(rotation_matrix)
+
+
+def create_rotation_2d(sin: Tensor, cos: Tensor) -> Tensor:
+    """
+    Create a 2d rotation matrix
+
+    Parameters
+    ----------
+    sin: Tensor
+        sin value to use for rotation matrix, [1]
+    cos: Tensor
+        cos value to use for rotation matrix, [1]
+
+    Returns
+    -------
+    Tensor
+        rotation matrix, [2, 2]
+    """
+    return torch.tensor([[cos.clone(), -sin.clone()], [sin.clone(), cos.clone()]])
+
+
+def create_rotation_3d(sin: Tensor, cos: Tensor) -> Tensor:
+    """
+    Create a 3d rotation matrix which sequentially applies the rotation
+    around axis (rot axis 0 -> rot axis 1 -> rot axis 2)
+
+    Parameters
+    ----------
+    sin: Tensor
+        sin values to use for the rotation, (axis 0, axis 1, axis 2)[3]
+    cos: Tensor
+        cos values to use for the rotation, (axis 0, axis 1, axis 2)[3]
+
+    Returns
+    -------
+    Tensor
+        rotation matrix, [3, 3]
+    """
+    rot_0 = create_rotation_3d_0(sin[0], cos[0])
+    rot_1 = create_rotation_3d_1(sin[1], cos[1])
+    rot_2 = create_rotation_3d_2(sin[2], cos[2])
+    return rot_2 @ (rot_1 @ rot_0)
+
+
+def create_rotation_3d_0(sin: Tensor, cos: Tensor) -> Tensor:
+    """
+    Create a rotation matrix around the zero-th axis
+
+    Parameters
+    ----------
+    sin: Tensor
+        sin value to use for rotation matrix, [1]
+    cos: Tensor
+        cos value to use for rotation matrix, [1]
+
+    Returns
+    -------
+    Tensor:
+        rotation matrix, [3, 3]
+    """
+    return torch.tensor([[1., 0., 0.],
```

Comment:
can we please provide dtype and device here? otherwise they may be created on CPU and with inappropriate types and we have to transfer them later, which is inefficient